File: src/marketplace_installer.py
"""Install / uninstall marketplaces by manipulating known_marketplaces.json.

A "marketplace install" is what Claude Code does when you run
`/plugins marketplace add owner/repo`:
 1. Clone the marketplace repo into ~/.claude/plugins/marketplaces/<name>/
 2. Register it in ~/.claude/plugins/known_marketplaces.json with installLocation
    and lastUpdated.

This module mirrors that behavior using the GitHub REST API (no git binary).
"""
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

from . import config
from .github_client import GitHubClient
from .installer import _atomic_write_json, _rmtree_robust


def known_marketplaces_dir() -> Path:
    return config.claude_plugins_dir() / "marketplaces"


def _load_known() -> dict:
    f = config.known_marketplaces_file()
    if not f.exists():
        return {}
    try:
        data = json.loads(f.read_text(encoding="utf-8"))
    except Exception:
        return {}
    return data if isinstance(data, dict) else {}


def _save_known(data: dict) -> None:
    _atomic_write_json(config.known_marketplaces_file(), data)


def _now_iso() -> str:
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def is_marketplace_installed(name: str) -> bool:
    return name in _load_known()


def get_install_info(name: str) -> dict:
    """Return the known_marketplaces.json record for a marketplace, or {}."""
    return _load_known().get(name, {}) or {}


def install_marketplace(gh: GitHubClient, name: str, repo: str, ref: str = "",
                        auto_update: bool | None = None) -> Path:
    """Download a marketplace repo and register it in known_marketplaces.json.

    `auto_update`:
      - True/False  → write the flag to the record.
      - None        → preserve the existing value, default False if no record yet.

    Raises ValueError if name/repo are missing, GitHubError on download failure.
    """
    if not name:
        raise ValueError("Marketplace name is required.")
    if not repo:
        raise ValueError(f"Marketplace '{name}' has no GitHub repo configured.")

    if not ref:
        ref = gh.get_default_branch(repo)

    install_path = known_marketplaces_dir() / name
    zip_bytes = gh.download_zipball(repo, ref=ref)
    _rmtree_robust(install_path)
    install_path.mkdir(parents=True, exist_ok=True)
    GitHubClient.extract_zipball(zip_bytes, install_path)

    sha = ""
    try:
        sha = (gh.get_latest_commit(repo, branch=ref) or {}).get("sha", "")
    except Exception:
        pass

    data = _load_known()
    existing = data.get(name) if isinstance(data.get(name), dict) else {}
    record = {
        "source": {"source": "github", "repo": repo},
        "installLocation": str(install_path),
        "lastUpdated": _now_iso(),
    }
    if sha:
        record["gitCommitSha"] = sha
    if auto_update is None:
        record["autoUpdate"] = bool(existing.get("autoUpdate", False))
    else:
        record["autoUpdate"] = bool(auto_update)
    data[name] = record
    _save_known(data)
    return install_path


def set_auto_update(name: str, value: bool) -> bool:
    """Update the `autoUpdate` flag in known_marketplaces.json for `name`.

    Returns False if the marketplace isn't installed (no record to patch).
    """
    data = _load_known()
    info = data.get(name)
    if not isinstance(info, dict):
        return False
    info["autoUpdate"] = bool(value)
    data[name] = info
    _save_known(data)
    return True


def auto_update_if_changed(gh: GitHubClient, name: str, repo: str, ref: str = "") -> tuple[bool, str]:
    """Re-install the marketplace if its remote HEAD differs from the stored SHA.

    Returns (updated, message). `updated=False` means "already up to date" or
    "no SHA recorded yet — skipping to avoid forced redownload" depending on
    `message`.
    """
    if not repo:
        return False, "no repo"
    if not ref:
        try:
            ref = gh.get_default_branch(repo)
        except Exception:
            ref = "main"
    info = get_install_info(name)
    stored = info.get("gitCommitSha", "") if isinstance(info, dict) else ""
    try:
        latest = (gh.get_latest_commit(repo, branch=ref) or {}).get("sha", "")
    except Exception as e:
        return False, f"check failed: {e}"
    if stored and latest and stored == latest:
        return False, "up to date"
    install_marketplace(gh, name, repo, ref=ref)
    return True, latest or "updated"


def uninstall_marketplace(name: str) -> None:
    """Remove the entry from known_marketplaces.json and delete the folder."""
    data = _load_known()
    info = data.pop(name, None)
    if info is not None:
        _save_known(data)
    install_path = (info or {}).get("installLocation") if isinstance(info, dict) else None
    if install_path:
        _rmtree_robust(Path(install_path))
    else:
        # Fallback: delete the conventional location if it exists
        _rmtree_robust(known_marketplaces_dir() / name)

File: src-tauri/src/marketplace_installer.rs
//! Install / uninstall marketplaces — port of src/marketplace_installer.py.

use crate::config;
use crate::error::{Error, Result};
use crate::github_client::GitHubClient;
use crate::installer::{atomic_write_json, now_iso, rmtree_robust};
use serde_json::{json, Map, Value};
use std::fs;
use std::path::PathBuf;

pub fn known_marketplaces_dir() -> PathBuf {
    config::claude_plugins_dir().join("marketplaces")
}

fn load_known() -> Map<String, Value> {
    let f = config::known_marketplaces_file();
    if !f.exists() {
        return Map::new();
    }
    fs::read_to_string(&f)
        .ok()
        .and_then(|s| serde_json::from_str::<Value>(&s).ok())
        .and_then(|v| v.as_object().cloned())
        .unwrap_or_default()
}

fn save_known(data: &Map<String, Value>) -> Result<()> {
    atomic_write_json(&config::known_marketplaces_file(), &Value::Object(data.clone()))
}

pub fn is_marketplace_installed(name: &str) -> bool {
    load_known().contains_key(name)
}

pub fn get_install_info(name: &str) -> Value {
    load_known().get(name).cloned().unwrap_or(Value::Null)
}

pub fn install_marketplace(
    gh: &GitHubClient,
    name: &str,
    repo: &str,
    r#ref: &str,
    auto_update: Option<bool>,
) -> Result<PathBuf> {
    if name.is_empty() {
        return Err(Error::Invalid("Marketplace name is required.".into()));
    }
    if repo.is_empty() {
        return Err(Error::Invalid(format!(
            "Marketplace '{name}' has no GitHub repo configured."
        )));
    }
    let r#ref = if r#ref.is_empty() {
        gh.get_default_branch(repo)?
    } else {
        r#ref.to_string()
    };
    let install_path = known_marketplaces_dir().join(name);
    let zip_bytes = gh.download_zipball(repo, &r#ref)?;
    rmtree_robust(&install_path)?;
    fs::create_dir_all(&install_path)?;
    GitHubClient::extract_zipball(&zip_bytes, &install_path, "")?;

    let sha = gh
        .get_latest_commit(repo, &r#ref)
        .ok()
        .and_then(|v| v.get("sha").and_then(|s| s.as_str()).map(String::from))
        .unwrap_or_default();

    let mut data = load_known();
    let existing = data
        .get(name)
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();
    let mut record = json!({
        "source": {"source": "github", "repo": repo},
        "installLocation": install_path.to_string_lossy(),
        "lastUpdated": now_iso(),
    });
    if !sha.is_empty() {
        record["gitCommitSha"] = json!(sha);
    }
    let auto = match auto_update {
        Some(v) => v,
        None => existing
            .get("autoUpdate")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
    };
    record["autoUpdate"] = json!(auto);
    data.insert(name.to_string(), record);
    save_known(&data)?;
    Ok(install_path)
}

pub fn set_auto_update(name: &str, value: bool) -> Result<bool> {
    let mut data = load_known();
    let Some(Value::Object(mut info)) = data.remove(name) else {
        return Ok(false);
    };
    info.insert("autoUpdate".into(), json!(value));
    data.insert(name.to_string(), Value::Object(info));
    save_known(&data)?;
    Ok(true)
}

pub fn auto_update_if_changed(
    gh: &GitHubClient,
    name: &str,
    repo: &str,
    r#ref: &str,
) -> (bool, String) {
    if repo.is_empty() {
        return (false, "no repo".into());
    }
    let r#ref = if r#ref.is_empty() {
        gh.get_default_branch(repo).unwrap_or_else(|_| "main".into())
    } else {
        r#ref.to_string()
    };
    let info = get_install_info(name);
    let stored = info
        .get("gitCommitSha")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let latest = match gh.get_latest_commit(repo, &r#ref) {
        Ok(v) => v
            .get("sha")
            .and_then(|s| s.as_str())
            .unwrap_or("")
            .to_string(),
        Err(e) => return (false, format!("check failed: {e}")),
    };
    if !stored.is_empty() && !latest.is_empty() && stored == latest {
        return (false, "up to date".into());
    }
    match install_marketplace(gh, name, repo, &r#ref, None) {
        Ok(_) => (true, if latest.is_empty() { "updated".into() } else { latest }),
        Err(e) => (false, format!("install failed: {e}")),
    }
}

pub fn uninstall_marketplace(name: &str) -> Result<()> {
    let mut data = load_known();
    let info = data.remove(name);
    if info.is_some() {
        save_known(&data)?;
    }
    let install_path = info
        .as_ref()
        .and_then(|v| v.get("installLocation"))
        .and_then(|v| v.as_str())
        .map(PathBuf::from);
    match install_path {
        Some(p) => {
            let _ = rmtree_robust(&p);
        }
        None => {
            let _ = rmtree_robust(&known_marketplaces_dir().join(name));
        }
    }
    Ok(())
}
